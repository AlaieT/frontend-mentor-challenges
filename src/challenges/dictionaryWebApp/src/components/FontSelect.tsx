import React from "react";

import type { FontSelectProps, Fonts } from "../types";

import { ReactComponent as ArrowDownIcon } from "../assets/images/icon-arrow-down.svg";

import styles from "../styles/components/fontSelect.module.scss";

const FontSelect = ({ callback = () => undefined, font, mode }: FontSelectProps) => {
  const [isSelectFont, setSelectFont] = React.useState(false);
  const handleSetFont = React.useCallback(
    (newFont: Fonts) => {
      setSelectFont(false);
      callback(newFont);
    },
    [callback]
  );

  return (
    <div id={styles.font} className={styles[mode]} onMouseLeave={() => setSelectFont(false)}>
      <div id={styles["current-font"]} onClick={() => setSelectFont(true)}>
        <p>{font}</p>
        <ArrowDownIcon width={12} height={6} />
      </div>
      {isSelectFont && (
        <ul
          id={styles["select-font"]}
        >
          <li
            className={styles.inter}
            onClick={() => handleSetFont("Sans Serif")}
          >
            Sans Serif
          </li>
          <li
            className={styles.lora}
            onClick={() => handleSetFont("Serif")}
          >
            Serif
          </li>
          <li
            className={styles.inconsolata}
            onClick={() => handleSetFont("Mono")}
          >
            Mono
          </li>
        </ul>
      )}
    </div>
  );
};

export default FontSelect;
