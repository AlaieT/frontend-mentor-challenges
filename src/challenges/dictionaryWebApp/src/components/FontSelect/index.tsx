import React from "react";

import { ReactComponent as ArrowDownIcon } from "../../assets/images/icon-arrow-down.svg";

import styles from "../../styles/components/fontSelect.module.scss";

const FontSelect = () => {
  const [currentFont, setFont] = React.useState("Sans Serif");
  const [isSelectFont, setSelectFont] = React.useState(false);

  return (
    <div id={styles.font}>
      <div id={styles["current-font"]} onClick={() => setSelectFont(true)}>
        <p>{currentFont}</p>
        <ArrowDownIcon width={12} height={6} />
      </div>
      <ul
        id={styles["select-font"]}
        hidden={!!isSelectFont}
        onMouseLeave={() => setSelectFont(false)}
      >
        <li onClick={() => setFont("Sans Serif")}>Sans Serif</li>
        <li onClick={() => setFont("Serif")}>Serif</li>
        <li onClick={() => setFont("Mono")}>Mono</li>
      </ul>
    </div>
  );
};

export default FontSelect;
