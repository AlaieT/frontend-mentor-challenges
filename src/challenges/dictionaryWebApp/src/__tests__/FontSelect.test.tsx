import React from "react";
import {
  render,
  cleanup,
  fireEvent,
  act,
  screen
} from "@testing-library/react";

import FontSelect from "../components/FontSelect";

afterEach(cleanup);

describe("FontsSelect", () => {
  describe("correct render", () => {
    it("should correct render without window for font selection", () => {
      expect(render(<FontSelect />).asFragment()).toMatchSnapshot();
    });

    it("should correct render with window for font selection", async () => {
      const { asFragment } = render(<FontSelect />);

      await act(async () => {
        fireEvent.click(
          (await screen.findAllByText(/sans serif/i))[0].parentElement
        );
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("correct functionality", () => {
    it("should correct change font type on clink", async () => {
      render(<FontSelect />);

      await act(async () => {
        fireEvent.click(
          (await screen.findAllByText(/sans serif/i))[0].parentElement
        );
      });

      await act(async () => {
        fireEvent.click((screen.getByText(/mono/i)));
      });      

      expect((await screen.findAllByText(/mono/i))[0]).toBeDefined();
    });
  });
});
