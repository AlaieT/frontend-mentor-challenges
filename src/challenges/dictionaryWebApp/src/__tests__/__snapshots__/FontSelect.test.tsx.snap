// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FontsSelect correct render should correct render with window for font selection 1`] = `
<DocumentFragment>
  <div
    id="font"
  >
    <button
      id="current-font"
    >
      <p>
        Sans Serif
      </p>
      <div
        height="6"
        width="12"
      />
    </button>
    <ul
      id="select-font"
    >
      <li>
        <button
          class="inter"
        >
          Sans Serif
        </button>
      </li>
      <li>
        <button
          class="lora"
        >
          Serif
        </button>
      </li>
      <li>
        <button
          class="inconsolata"
        >
          Mono
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`FontsSelect correct render should correct render without window for font selection 1`] = `
<DocumentFragment>
  <div
    id="font"
  >
    <button
      id="current-font"
    >
      <p>
        Sans Serif
      </p>
      <div
        height="6"
        width="12"
      />
    </button>
  </div>
</DocumentFragment>
`;
